Added lots of DB search queries

silent-po2 opened the pull request:
Specifically added queries for author name, year, material (composition or details supported), category ID, subcategory ID, and any combination of these, plus a query based on a dataset ID.

Currently, this must be run/tested against the local DB, as the remote DB tables are currently missing some key data. This will be remedied in the near future.

GD-Dheer committed:
[#71] Backend logic to parse CSV file

GD-Dheer committed:
[#71] Backend logic to parse CSV file

GD-Dheer committed:
[#71] Logic for sending CSV to backend + light cleanup

GD-Dheer committed:
[#71] Update backend to support frontend changes

laithAwad committed:
#82: Auto stash before merge of "hiding-dataset-#82" and "origin/Colour-scheme-issue-#99"

laithAwad committed:
#82: Auto stash before merge of "hiding-dataset-#82" and "origin/Colour-scheme-issue-#99"

laithAwad committed:
Hiding Datasets

Added a button that can hide specific datasets from the graph

Commit:
Added toggle between linear scale and log scale.

Added two circles on the axes which toggle between linear scales and log scales.

Commit:
Clean up code

Commit:
Refactored Scale Toggle

Swapped the scale toggle from being shapes to being actual buttons.

Commit:
Added comments

laithAwad committed:
Trying to hide datasets using legends

GD-Dheer committed:
[#71] Add routing for upload page

GD-Dheer committed:
[#71] Temp changes for JSON support

GD-Dheer committed:
[#71] Support to upload & process JSON

GD-Dheer committed:
[#71] Fix comments

laithAwad committed:
Added hide dataset option

Hiding datasets using the legend dots and showing them on click

laithAwad committed:
Added hide dataset option

Hiding datasets using the legend dots and showing them on click

silent-po2 committed:
[#53] Created initial queries for querying data set.
Also had to add a "None Inputted" value for Representations as having those be null in data points was causing a query to fail

silent-po2 committed:
[#53] Created initial queries for querying data set.
Also had to add a "None Inputted" value for Representations as having those be null in data points was causing a query to fail

silent-po2 committed:
[#53] Created initial queries for querying data set.
Also had to add a "None Inputted" value for Representations as having those be null in data points was causing a query to fail

silent-po2 committed:
[#53] Created initial queries for querying data set.
Also had to add a "None Inputted" value for Representations as having those be null in data points was causing a query to fail

laithAwad committed:
Finished #82

Hiding dataset completely when clicked on the eye icon inside the legend

laithAwad committed:
Finished #82

Hiding dataset completely when clicked on the eye icon inside the legend

Vatika1 committed:
removing dataset

Vatika1 committed:
removing dataset

Commit:
Empty Commit

Commit:
Empty Commit

Vatika1 committed:
Created button to delete datasets

included drop-down based on datasets and selected data

Vatika1 committed:
Created button to delete datasets

included drop-down based on datasets and selected data

Vatika1 committed:
Correction: selenium setup on Client directory

GD-Dheer committed:
[#71] Add Model & process half contents

GD-Dheer committed:
[#71] Add Model & process half contents

GD-Dheer committed:
[#71] Re-add styled-components into package.json

Commit:
Remove comments

silent-po2 committed:
#53 initial WIP queries. This show the potential to get every single data set plus individual values from the RowDataPackets

silent-po2 committed:
#53 initial WIP queries. This show the potential to get every single data set plus individual values from the RowDataPackets

silent-po2 committed:
#53 initial WIP queries. This show the potential to get every single data set plus individual values from the RowDataPackets

GD-Dheer committed:
[#71] Update models with almost final changes

GD-Dheer committed:
[#71] Add dataSetComments

TiraFire committed:
Delete dataset updates ID list now

The delete dataset did not update the new id list, so it does that now, and updates the delete dataset options as well.

GD-Dheer committed:
[#71] Fix css for upload

GD-Dheer committed:
[#71] Update package.json

GD-Dheer committed:
[#71] Update package.json

GD-Dheer committed:
[#71] Fix spacing

silent-po2 committed:
[#71] Did some minor DB updates to add some default values to quash potential errors in new data set uploading

silent-po2 committed:
[#71] Added instantiated DB to ensure fallbacks function as intended

silent-po2 committed:
[#53] Queries can now work based on a variable dataset id instead of a fixed one

silent-po2 committed:
#53 added other queries to cover all the data of both data sets

silent-po2 committed:
#53 added other queries to cover all the data of both data sets

silent-po2 committed:
[#53] Fixed some oddities resulting from merging latest develop into branch

silent-po2 committed:
[#53] Fixed some oddities resulting from merging latest develop into branch

silent-po2 committed:
[#53] Additional query clean up

silent-po2 committed:
[#53] Additional query clean up

GD-Dheer committed:
[53] Add router, test branch commit

GD-Dheer committed:
[53] Add router, test branch commit

GD-Dheer committed:
[#53] Add routing to app

GD-Dheer committed:
[#53] Add routing to app

GD-Dheer committed:
[#53] Add getData controller & service

GD-Dheer committed:
[#53] Add getData controller & service

GD-Dheer committed:
[#53] Add Model for querying

GD-Dheer committed:
[#53] Add Model for querying

GD-Dheer committed:
[#53] Fix merge conflicts

silent-po2 committed:
[#53] Re-added queries

silent-po2 committed:
[#53] undoing refill of database.ts; this was done in error

silent-po2 committed:
[#53] Added search queries for any of the following: data set ID (already present), material (accepts composition or material details), author (needs first and last name), publication year, category id, and category id+subcategory id. Each query returns a 2D array to the front end which has everything that
Additionally cleared out some now unused /routes, /models, and /controllers files from prior work on this issue

silent-po2 committed:
[#53] Added search queries for any of the following: data set ID (already present), material (accepts composition or material details), author (needs first and last name), publication year, category id, and category id+subcategory id. Each query returns a 2D array to the front end which has everything that
Additionally cleared out some now unused /routes, /models, and /controllers files from prior work on this issue

silent-po2 committed:
[#53] Bumping branch up in 'active branch' list via newline addition

silent-po2 committed:
[#53] Bumping branch up in 'active branch' list via newline addition

silent-po2 committed:
[#53] Attempted alteration of returned data from a 2D array with a new IDatasetResponseModel

silent-po2 committed:
[#53] Attempted alteration of returned data from a 2D array with a new IDatasetResponseModel

silent-po2 committed:
[#53] All returned data now correctly utilizes the IDatasetResponseModel. Queries now accept params rather than fixed values. Refactored a number of SelectQueryDatabase code to be cleaner, will move segments of this new code to other Models as appropriate in the next commit

silent-po2 committed:
[#53] All returned data now correctly utilizes the IDatasetResponseModel. Queries now accept params rather than fixed values. Refactored a number of SelectQueryDatabase code to be cleaner, will move segments of this new code to other Models as appropriate in the next commit

silent-po2 committed:
[#53] All search queries now use IDatasetResponseModel as intended. Fixed issues where sub-queries of finding all materials based on one material and finding all authors based on one author did not work.

silent-po2 committed:
[#53] All search queries now use IDatasetResponseModel as intended. Fixed issues where sub-queries of finding all materials based on one material and finding all authors based on one author did not work.

silent-po2 committed:
[#53] Some minor formatting tweaks

silent-po2 committed:
[#53] Some minor formatting tweaks

silent-po2 committed:
[#53] Added all possible query calls to getDataService and did some minor changes to SelectQueryDatabase

silent-po2 committed:
[#53] Added all possible query calls to getDataService and did some minor changes to SelectQueryDatabase

silent-po2 committed:
[#53] Tweaked material and author WHERE statements in SelectQueryDatabase.ts

silent-po2 committed:
[#53] Tweaked material and author WHERE statements in SelectQueryDatabase.ts

silent-po2 committed:
[#53] Added query for Material+Author case. Note that getDataFromMaterialYearAuthorSubcategory is not complete, is a WIP

silent-po2 committed:
[#53] Added query for Material+Author case. Note that getDataFromMaterialYearAuthorSubcategory is not complete, is a WIP

silent-po2 committed:
[#53] Added Material+Author+Year query

silent-po2 committed:
[#53] Added Material+Author+Year query

silent-po2 committed:
[#53] getDataFromMaterialYearAuthorSubcategory query complete

silent-po2 committed:
[#53] getDataFromMaterialYearAuthorSubcategory query complete

silent-po2 committed:
[#53] Material+Author+Year+Category query added

silent-po2 committed:
[#53] Material+Author+Year+Category query added

silent-po2 committed:
[#53] Added getDataFromMaterialAuthorSubcategory

silent-po2 committed:
[#53] Added getDataFromMaterialAuthorSubcategory

silent-po2 committed:
[#53] added getDataFromMaterialAuthorCategory

silent-po2 committed:
[#53] added getDataFromMaterialAuthorCategory

silent-po2 committed:
[#53] Added getDataFromMaterialYearSubcategory

silent-po2 committed:
[#53] Added getDataFromMaterialYearSubcategory

silent-po2 committed:
[#53] added getDataFromMaterialYearCategory

silent-po2 committed:
[#53] added getDataFromMaterialYearCategory

silent-po2 committed:
[#53] Fixed some error causing typos

silent-po2 committed:
[#53] Fixed some error causing typos

silent-po2 committed:
[#53] Added getDataFromMaterialSubcategory, getDataFromMaterialCategory

silent-po2 committed:
[#53] Added getDataFromMaterialSubcategory, getDataFromMaterialCategory

silent-po2 committed:
[#53] added getDataFromMaterialYear

silent-po2 committed:
[#53] added getDataFromMaterialYear

silent-po2 committed:
[#53] added getDataFromYearSubcategory, getDataFromYearCategory

silent-po2 committed:
[#53] added getDataFromYearSubcategory, getDataFromYearCategory

silent-po2 committed:
[#53] Completed all queries

silent-po2 committed:
[#53] Completed all queries

silent-po2 committed:
[#53] Reverting two files to develop branch state to prevent merge conflict issue

silent-po2 committed:
[#53] Reverting two files to develop branch state to prevent merge conflict issue

silent-po2 committed:
[#53] Fixed some merge conflicts with develop

silent-po2 committed:
[#53] More merge conflict fixing

GD-Dheer commented on Client/package.json:
```diff
@@ -16,7 +16,8 @@
     "react-scripts": "3.4.3",
-    "selenium-webdriver": "^4.0.0-alpha.7"
+    "selenium-webdriver": "^4.0.0-alpha.7",
+    "styled-components": "^5.2.1"
```
This was removed in develop

silent-po2 replied:
Yes, this is an unintentional merge conflict. Will fix

GD-Dheer commented on Server/src/controllers/getDataController.ts:
```diff
@@ -0,0 +1,11 @@
+const getDataService = require('../services/getDataService');
```
This is all still following javascript structure not the new TS structure. Needs to be fixed - see Refactoring to TS

silent-po2 replied:
Should be fixed now

GD-Dheer commented on Server/src/database.ts:
```diff
@@ -16,5 +16,168 @@ export const connectDB = async () => {
+      .innerJoin(Dataset, 'dataset', 'datapointcomments.datasetId = dataset.id')
+      .where('dataset.id = :datasetId', { datasetId: dataset2 })
+      .getRawMany();
+    console.log(datapointComments2);
```
Are all of these queries needed? I dont think any of these belong in db.ts

silent-po2 replied:
This is an unintentional merge conflict. These queries shouldn't be there; will fix

GD-Dheer commented on Server/src/models/SelectQueryDatabase.ts:
```diff
@@ -0,0 +1,2806 @@
+import { EntityManager, getConnection } from "typeorm";
```
This file needs to be split into maybe 2-3 smaller ones. Debugging 2000 line files can get tricky very fast

GD-Dheer commented on Server/src/models/SelectQueryDatabase.ts:
```diff
@@ -0,0 +1,2806 @@
+    dataset: IDatasetModel[]
+    materials: IMaterialModel[]
+    dataPoints: IDataPointModel[]
+    dataPointComments: IDataPointCommentModel[]
```
Nice use of interfaces here

GD-Dheer commented on Server/src/models/SelectQueryDatabase.ts:
```diff
@@ -0,0 +1,2806 @@
+//         return this.andWhere('dataset.id = :datasetId', { datasetId: id })
+//     }
+// }
+
```
TS comment applies here

GD-Dheer commented on Server/src/models/SelectQueryDatabase.ts:
```diff
@@ -0,0 +1,2806 @@
+        dataPoints: subcategoryDatapointData,
+        dataset: subcategoryDatasetData,
+        materials: subcategoryMaterialData,
+        dataPointComments: subcategoryDatapointComments
```
There is tons of good queries here but working through this is impossible. We should group common queries together

GD-Dheer commented on Server/src/routes/queryDataRouter.ts:
```diff
@@ -0,0 +1,13 @@
+var express = require('express');
+const getDataRouter = require('express').Router();
+const getDataController = require('../controllers/getDataController');
```
TS comment applies here

silent-po2 replied:
Should be fixed now

GD-Dheer commented on Server/src/routes/queryDataRouter.ts:
```diff
@@ -0,0 +1,13 @@
+ * to continue processing of the request.
+ */
+
+getDataRouter.get('/dataset*', getDataController.createRequestForData);
```
We should do this a little cleaner than wild card here. Im pretty sure this lets the user submit anything at all

silent-po2 replied:
The user can submit anything, but it is all filtered by getDataService. Something cleaner should be possible, as we know exactly what sort of params a user can send, as per the list below:

datasetId
material
firstName
lastName
year
categoryId 
subcategoryId

That said, I'm not entirely sure how to approach such a change outside of making 24 routes for each possible query, which is equally messy.

GD-Dheer replied:
We dont need 24 routes. We can make them all optional parameters but we should specify the accepted parameters

silent-po2 replied:
That would work well, but I'm not sure where to start for that code wise. Do you have some documentation or other links for such?

GD-Dheer replied:
I realize with express we don't really need to. Since we have access to all input params with request.query.{paramName} so I suppose its ok to leave it as such

GD-Dheer commented on Server/src/services/getDataService.ts:
```diff
@@ -0,0 +1,122 @@
+import { EntityManager } from "typeorm";
+
+const obtainDataModel = require('../models/SelectQueryDatabase');
```
TS comment applies here

GD-Dheer commented on Server/src/services/getDataService.ts:
```diff
@@ -0,0 +1,122 @@
+    if (datasetReceived != undefined) {
+        setOfData = await obtainDataModel.getDataFromDataset(datasetReceived);
+    }
+    else if (materialReceived != undefined) {
```
Lets avoid these if statements. Switch statements are better.... Even better - https://refactoring.guru/replace-conditional-with-polymorphism. Follows the strategy pattern with polymorphism. I can assist with this if its required.

GD-Dheer reviewed (dismissed):
A+ work. Doing all of this was not easy. This handles all of the functionality of the old website.

silent-po2 committed:
[#53] Last fix of merge conflicts

Lavo-2 committed:
[#53] TS formatting changes implemented

Lavo-2 committed:
[#53] Minor fix for a smell in Datapoints.ts

ds161833 commented:
Needs modification to return an array of datasets

Lavo-2 committed:
[#53] minor refactor tweak

Lavo-2 committed:
[#53] Split off interfaces from SelectQueryDatabase.ts into their own file

Lavo-2 committed:
[#53] Split selectQueryDatabase contents among 3 different files for future ease of debugging

silent-po2 committed:
[#53] Major refactor in progress, once completed will reduce lines of code by at least 1000

GD-Dheer commented on Server/src/models/SelectQuerySingleTerm.ts:
```diff
@@ -421,4 +426,186 @@ export const getDataFromSubcategory = async (category: number, subcategory: numb
+    return subcategoryDatasetData;
+}
+
+export const selectAuthorsQuery2 = (manager: EntityManager, idArray) =>
```
Better names please!

silent-po2 committed:
[#53] Refactoring complete

silent-po2 committed:
[#53] Merge branch 'develop' into backend/#53-query-data-sets.
Additionally set DB to use local DB instead of remote DB for the time being as queries will not work on the remote DB as the remote DB tables are incomplete

silent-po2 committed:
[#53] Returned set of data now returns an array of IDatasetResponseModel models with each array entry being a complete data set

silent-po2 committed:
[#53] Merge branch 'develop' of https://github.com/Data-Boom/SOEN490 into backend/#53-query-data-sets

silent-po2 commented:
This code has been entirely refactored and brought down from over 2600 lines of code to around 400, in large part thanks to a great idea from @TiraFire this morning. 

> Needs modification to return an array of datasets

As of commit 804f6f0 this is now done, thanks for the feedback!

silent-po2 committed:
[#53] Minor refactoring and made changes to enable full functionality with remote DB. Additionally fixed the category/subcategory bug mentioned in #150

silent-po2 committed:
[#53] Minor refactor

silent-po2 committed:
[#53] More minor refactoring

silent-po2 committed:
[#53] Removed some no longer needed console logs

silent-po2 committed:
[#53] Renamed a variable for better clarity

silent-po2 committed:
[#53] Removed /Database folder as it is no longer needed

silent-po2 committed:
[#53] Refactored for loop in getDataService.ts into a dedicated class

GD-Dheer commented on Server/src/services/getDataService.ts:
```diff
@@ -79,14 +73,20 @@ export const getDatasetIdsFromParams = async (materialReceived: string, yearRece
     return selectedDatasetIds
 }
 
+export const createDatasetIdArray = async (rawData: any[]) => {
```
this is a refactor? How?

GD-Dheer commented on Server/src/loaders/loadStartupProcess.ts:
```diff
@@ -5,6 +5,7 @@ import { connectDB } from '../database'
 import cors from 'cors';
 import express from 'express';
 import { fileUploadRouter } from '../routes/fileUploadRouter'
+import { getDataRouter } from '../routes/queryDataRouter';
```
I think we should try to be careful with what we consider to be refactoring... Ideally its when we restructure our codebase to make it more optimal and/or more flexible

silent-po2 replied:
You're right, I was misusing the term here. It won't happen in the future.

GD-Dheer commented:
> This code has been entirely refactored and brought down from over 2600 lines of code to around 400, in large part thanks to a great idea from @TiraFire this morning.
> 
> > Needs modification to return an array of datasets
> 
> As of commit [804f6f0](https://github.com/Data-Boom/SOEN490/commit/804f6f079d3ece95bdee574c0eb01a41b9fd3f74) this is now done, thanks for the feedback!

A++

silent-po2 committed:
[#53] TS conversion should be completed now

GD-Dheer commented on Server/src/controllers/getDataController.ts:
```diff
@@ -0,0 +1,12 @@
+import { retrieveData } from '../services/getDataService';
+
+
+export const createRequestForData = async (req, res) => {
```
Still needs to be converted to class

silent-po2 replied:
Good catch, rectified this in commit 3c228c5

silent-po2 committed:
[#53] Missed TS conversion of getDataController. That has been rectified

silent-po2 committed:
[#53] Minor tweak

TiraFire commented on Server/src/models/DataQueryModel.ts:
```diff
@@ -0,0 +1,124 @@
+    async getAllData(id: number): Promise<IDatasetResponseModel> {
+
+        const connection = getConnection();
+        console.log("Getting all data sets' raw data");
+
+        console.log("Getting publications");
+        let publicationData: IPublicationModel[] = await selectPublicationsQuery(connection.manager, id)
+
+        console.log("Getting authors");
+        let authorData: IAuthorModel[] = await selectAuthorsQuery(connection.manager, id)
+
+        console.log("Getting data sets");
+        let datasetData: IDatasetModel[] = await selectDatasetsQuery(connection.manager, id)
+
+        console.log("Getting data point data");
+        let datapointData: IDataPointModel[] = await selectDataPointsQuery(connection.manager, id)
+
+        console.log("Getting data set materials");
+        let materialData: IMaterialModel[] = await selectMaterialQuery(connection.manager, id)
+
+        console.log("Getting data point comments");
```
Will the console logs remain or will they be removed and replaced by comments? Are they for testing or another reason?

silent-po2 replied:
They were for testing, I will remove them shortly

silent-po2 replied:
Done in commit 6066590

TiraFire commented on Server/src/models/entities/Dataset.ts:
```diff
@@ -77,4 +77,22 @@ export class Dataset {
+    manager.createQueryBuilder(Dataset, 'dataset')
+        .select('dataset.id', 'dataset_id')
+
+export const selectDatasetsQuery = (manager: EntityManager, dataset: number) =>
```
Good work on the queries

TiraFire commented on Server/src/models/interfaces/DataRequestModelInterface.ts:
```diff
@@ -0,0 +1,9 @@
+export interface IDataRequestModel {
```
I like this model for data requests, if ranges are included or multiple choices for material, how will those be handled?

silent-po2 replied:
Will talk with frontend to figure out how many materials they want maximum in a search, this can be expanded.

silent-po2 replied:
As of commit 529a1d5 the GET request now accepts an array of materials, allowing for multi-material search functionality.

TiraFire commented on Server/src/services/getDataService.ts:
```diff
@@ -0,0 +1,110 @@
+        let categoryReceived = receivedData.categoryId;
+        let subcategoryReceived = receivedData.subcategoryId;
+        let selectedDatasetIds = [];
+        if (datasetReceived != undefined) {
```
This can be removed with the selectedDatasetIds already set, and if for undefined (changes we implemented during our chat)

silent-po2 replied:
Done in commit 6a0eac9

TiraFire commented on Server/src/services/getDataService.ts:
```diff
@@ -0,0 +1,110 @@
+        let currentDatasetId: number
+        for (let i = 0; i < allDatasetIds.length; i++) {
+            count = 1;
+            currentDatasetId = allDatasetIds[i]
```
No need for this, just use the allDatasetIds[i] (as mentioned in the chat)

silent-po2 replied:
Done in commit 6a0eac9

silent-po2 committed:
[#53] Minor tweak

silent-po2 committed:
[#53] Removed no longer needed console logs

silent-po2 committed:
[#53] Minor tweak

silent-po2 committed:
[#53] Minor tweak

silent-po2 committed:
[#53] GET request will now accept an array of materials, entered in the following format: material[]=X&material[]=Y&material[]=Z&...material[]=N
Where X, Y, Z, N are a material that one wishes to search for.
The order of the material[] calls in the get does not matter. The GET below works without issue for example:
/dataset?material[]=carbon, graphite, pressed graphite&year=1980&material[]=O2

JeremiahTee commented:
Great work with this PR. I went over every change with Sean during a peer programming session & recorded his demo. It is up on Google drive for the team to review for knowledge sharing. Postman was used during the demo and Sean clearly proved that all the mentioned queries are working.

As well, he added support for multiple materials on 529a1d5. I apologize for approving this now, this should've been approved right after the demo but there was some miscommunication.

The miscommunication being that it was decided that unit tests for this issue will be done in a separate PR - as this user story is a blocker for some members.
